private/buf/cmd/buf/command/format/format.go

```diff
@@ -242,10 +252,62 @@ func run(
 	if flags.Output != "-" && flags.Write {
 		return fmt.Errorf("--%s cannot be used with --%s", outputFlagName, writeFlagName)
 	}
+	if flags.Stdin && flags.Output != "-" {
+		return fmt.Errorf("--%s cannot be used with --%s", stdinFlagName, outputFlagName)
+	}
+	if flags.Stdin && flags.Write {
+		return fmt.Errorf("--%s cannot be used with --%s", stdinFlagName, writeFlagName)
+	}
 	input, err := bufcli.GetInputValue(container, flags.InputHashtag, ".")
 	if err != nil {
 		return err
 	}
+	if flags.Stdin && container.NumArgs() > 0 {
+		return fmt.Errorf("--%s cannot be used with a non-empty input %s", stdinFlagName, input)
+	}
+	runner := command.NewRunner()
+	if flags.Stdin {
+		// We need to read the raw bytes from the io.Reader upfront so that
+		// the content can be referenced more than once: once by the formatter,
+		// and again during the diff.
```

Comment:
I smell something here. `--stdin` is implemented with a divergent different code path because its input is a non-seekable file. Standard input is not the only non-seekable file out there: sockets, tape-archive on literal tape, named pipes, bash's named descriptors (/dev/fd/<n>), etc.

One alternative is to implement a buffered, seeking reader. Wrap stdin with your reader and no more special code paths, I hope. I haven't groked all of the format package yet and `run` is particularly hard to follow.

This concept can extend further if you open a file input and attempt to seek to offset 0. That should, on POSIX systems, fail on pipes, sockets, and other file handles that are non-seekable. Commands like `buf format <(cat foo.proto bar.proto)` become possible.